### ISSUE TITLE
Allowing multiple authentication flows in parallel (e.g., in different tabs)

### ISSUE THREAD

**Issue by dmalan** [lepture/authlib#285]
**Is your feature request related to a problem? Please describe.**

We currently use `authlib.integrations.flask_client` to authenticate users against Auth0, where they can log in via different connection types (e.g., GitHub, Google, etc.). But users sometimes initiate multiple authentication flows in parallel, as by command- or control-clicking multiple links, each of which opens in a new tab. Even if the user logs in successfully to each of those tabs, each tab (except for the most recently opened one) ultimately errs with:

```
authlib.integrations.base_client.errors.MismatchingStateError: 
mismatching_state: CSRF Warning! State not equal in request and response
```

That seems to be the result of Authlib using a single session key (e.g., `_name_authlib_state_`) to implement CSRF protection, per https://github.com/lepture/authlib/blob/master/authlib/integrations/flask_client/integration.py#L15-L21.

Because each tab has a new value for `state`, previously opened tabs' state values are clobbered in Flask's session.

**Describe the solution you'd like**

Might it make sense to use unique session keys instead so that tabs don't share (and clobber each other's) state? Similarly for `_name_authlib_redirect_uri_` and `_name_authlib_nonce_`? E.g., instead of, in pseudocode,

```py
session["_name_authlib_state_"] = "ABC"
session["_name_authlib_redirect_uri_"] = "https://example.com/callback"
session["_name_authlib_nonce_"] = "XYZ"
```

perhaps an approach along these lines, whereby the key includes the unique `state` value?

```py
session["_name_authlib_ABC_"] = {
    "redirect_uri": "https://example.com/callback",
    "nonce": "XYZ"
}
```

**Describe alternatives you've considered**

Instead of `authlib.integrations.flask_client`, should we perhaps use `requests_client.OAuth2Session` instead to achieve this behavior? Would we need to validate `state`, `redirect_uri`, and `nonce` ourselves, though, and also validate the ID token returned from Auth0? 

Is a simpler approach already possible perhaps?

**Additional context**

If a user opens multiple tabs because they've command- or control-clicked multiple links (each of which requires authentication), it's indeed intentional that we want them to authenticate in each of those tabs (especially since their choice of connection type on Auth0's end might differ per tab). We wouldn't want one tab, upon encountering a `MismatchingStateError`, simply to check whether the user has logged in via another tab and allow the user to proceed.

Thank you!

**Comment by lepture** [comment]
Do you want to keep each tab logged in as different user? For instance:

Tab 1: logged in with Google as User A.
Tab 2: logged in with Twitter as User B.

Do you mean this? This is not possible.

If you just want to avoid `MismatchingStateError`, you can check if a user is logged in already in the authorization callback:

```
@app.route('/auth')
def authorize_callback():
    if current_user:
        return redirect('/')
    token = oauth.auth0.authorize_access_token()
    ....
    return '...'
```

**Comment by dmalan** [comment]
Not quite logged in as a different user in each tab, but authenticated independently in each tab. For instance, if we have links like

```
<a href="https://www.example.com/foo">foo</a>
<a href="https://www.example.com/bar">bar</a>
<a href="https://www.example.com/baz">baz</a>
```

each of which leads to a protected resource that requires authentication, it's not uncommon for users to open all three tabs at once, as by command- or control-clicking on all three, to work their way through them one at a time. But we want them to have to authenticate in each tab, not just one. (It's a deliberate UX decision, so that users have to agree to certain terms before viewing each resource.)

Is there perhaps a way to suppress the library's own CSRF-checking, so that we can do it ourselves with our own cookies? 

Or should we perhaps register a new Authlib client (using the same IdP) for each authentication flow (i.e., tab), as with code like the below, using the value of `state` as the client's "name" so that each flow is independent? In our callback, we could then call `oauth[state].authorize_access_token()` so as to use the right client when getting the access token, and then deregister the client somehow?

```
oauth = OAuth(app)

state = authlib.common.security.generate_token()
oauth.register(
    state,
    client_id=os.environ.get("CLIENT_ID"),
    client_kwargs={"scope": "openid profile email"},
    client_secret=os.environ.get("CLIENT_SECRET"),
    server_metadata_url=os.environ.get("SERVER_METADATA_URL")
)
```

Many thanks!

**Comment by lepture** [comment]
Here is my thoughts:

1. In tab 1, user is logged in with Google, you mark the end user is logged in as User A.
2. In tab 2, user is logged in with Twitter, since you have marked the end user is logged in, before you `authorize_access_token`, you know this browser is already authenticated, you just ignore this callback, don't call `authorize_access_token`, instead, you make a `302` redirection to refresh the page.

**Comment by dmalan** [comment]
Thanks for the follow-up! Afraid this approach doesn't satisfy our requirement that the user has to authenticate in both tabs, so that they have to agree to certain terms before viewing each resource. And it can still yield an `MismatchingStateError` exception if the user:

1. Opens https://www.example.com/foo in tab 1 but doesn't yet log in.
1. Opens https://www.example.com/bar in tab 2 but doesn't yet log in.
1. Returns to tab 1 and tries to log in.

At that point, they encounter a `MismatchingStateError`, because step #2 clobbers the value of `session["_name_authlib_state_"]` with a new state value. Our own UX goal aside, that seems to be the source of the issue, since the same cookie name is used for all outstanding authentication flows for a given client? But adding the actual value of `state` to the session *key* would seem to avoid perhaps, a la the below?

```py
session["_name_authlib_ABC_"] = {
    "redirect_uri": "https://example.com/callback",
    "nonce": "XYZ"
}
```

Thanks!

**Comment by lepture** [comment]
How could that be? When opening tab 2, why should `session["_name_authlib_state_"]` be changed? Is there a sample code?

**Comment by dmalan** [comment]
Sure, here's a sample app, https://github.com/cs50/id/. Runnable with:

```
cd flask
CLIENT_ID=... CLIENT_SECRET=... SERVER_METADATA_URL=... flask run
```

Reproducible with:

1. Visit http://localhost:5000/ in tab 1. Click **login** but don't log into idP.
1. Visit http://localhost:5000/ in tab 2. Click **login** but don't log into idP.
1. Return to tab 1. Finish logging in.

That'll yield an HTTP 500 with a `MismatchingStateError`:

```
ERROR in app: Exception on /callback [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/mnt/flask/application.py", line 51, in callback
    token = oauth.cs50.authorize_access_token()
  File "/usr/local/lib/python3.7/site-packages/authlib/integrations/flask_client/remote_app.py", line 74, in authorize_access_token
    params = self.retrieve_access_token_params(flask_req, request_token)
  File "/usr/local/lib/python3.7/site-packages/authlib/integrations/base_client/base_app.py", line 144, in retrieve_access_token_params
    params = self._retrieve_oauth2_access_token_params(request, params)
  File "/usr/local/lib/python3.7/site-packages/authlib/integrations/base_client/base_app.py", line 125, in _retrieve_oauth2_access_token_params
    raise MismatchingStateError()
authlib.integrations.base_client.errors.MismatchingStateError: mismatching_state: CSRF Warning! State not equal in request and response.
```

The reason seems to be because `oauth.cs50.authorize_redirect ` returns a URL like 

```
https://EXAMPLE.auth0.com/authorize?response_type=code&client_id=...&redirect_uri=http%3A%2F%2Flocalhost%3A5000%2Fcallback&scope=openid+profile+email&state=ABC&nonce=XYZ
```

in tab 1 and a URL like

```
https://EXAMPLE.auth0.com/authorize?response_type=code&client_id=...&redirect_uri=http%3A%2F%2Flocalhost%3A5000%2Fcallback&scope=openid+profile+email&state=123&nonce=456
```

in tab 2. The first call results in the library setting


```py
session["_cs50_authlib_state_"] = "ABC"
```

but the second call results in the library setting


```py
session["_cs50_authlib_state_"] = "123"
```

thereby clobbering the value of `session["_cs50_authlib_state_"]`, effectively invalidating tab 1, even if the flow hasn't yet been completed. But that could be avoided by not reusing the same session key for every authentication request? I.e., adding the value of `state` to the session *key* itself could allow each tab to be independent and also allow for CSRF checking, a la the below?

session["_cs50_authlib_ABC_"] = {
    "redirect_uri": "http://localhost:5000/callback",
    "nonce": "XYZ"
}

Thanks!

**Comment by lepture** [comment]
Oh, got it. When using `@login_required` on `/foo` and `/bar`, that will happen:

1. visit `/foo` will redirect to `/login`
2. visit `/bar` will redirect to `/login`
3. `/login` will redirect with `oauth.authorize_redirect`

But usually, `/login` will be a normal page.

**Comment by lepture** [comment]
This should be fixed by https://github.com/lepture/authlib/commit/2af1b9f2c81d5cb847491e96cf762f69215fcf2d

@dmalan can you verify with the latest master code?

**Comment by dmalan** [comment]
Hi @lepture, afraid the issue remains on `master`, reproducible via these same steps, https://github.com/lepture/authlib/issues/285#issuecomment-717767791, which still yield a `mismatching_state` error. The source of the issue still seems to be that each call to `authorize_redirect` results in global session keys being set for a given idP. E.g.:

1. `_cs50_authlib_redirect_uri_`
1. `_cs50_authlib_state_`
1.  `_cs50_authlib_nonce_`

The result is that if a user opens two tabs, both of which requires authentication, resulting in `authorize_redirect` being called twice, first for tab #1, then for tab #2, the second call to `authorize_redirect` clobbers the global session values. For instance, if you add `print(session)` after any call to `authorize_redirect`, you might see this in tab #164 

```
<FileSystemSession {'_cs50_authlib_redirect_uri_': 'http://localhost:8080/redirect_uri', '_cs50_authlib_state_': '72HQmRFEObIzistjSQOrYc79K9so0p', '_cs50_authlib_nonce_': 'iAWZwBXzCs8iVj6l4A6N'}>
```

and this in tab #2:

```
<FileSystemSession {'_cs50_authlib_redirect_uri_': 'http://localhost:8080/redirect_uri', '_cs50_authlib_state_': 'FS1lbk8e8nMMZmeS2Smh3DakQtkdY4', '_cs50_authlib_nonce_': 'iAWZwBXzCs8iVj6l4A6N'}>
```

Note that the value for `_cs50_authlib_state_` has changed. The result is that tab #1's state is effectively invalidated, so if the user tries to complete the authentication flow in tab #1, they end up with a `mismatching_state` error.

The ideal fix would seem to be not to use idP-specific session keys (e.g., `_cs50_authlib_redirect_uri_`, `_cs50_authlib_state_`, and `_cs50_authlib_nonce_`) but instead use state-specific session keys? E.g., tab #1's session keys could be:

```
<FileSystemSession {'_cs50_authlib_72HQmRFEObIzistjSQOrYc79K9so0p_redirect_uri_': 'http://localhost:8080/redirect_uri', '_cs50_authlib_72HQmRFEObIzistjSQOrYc79K9so0p_nonce_': 'iAWZwBXzCs8iVj6l4A6N'}>
```

and this in tab #2:

```
<FileSystemSession {'_cs50_authlib_FS1lbk8e8nMMZmeS2Smh3DakQtkdY4_redirect_uri_': 'http://localhost:8080/redirect_uri', '_cs50_authlib_FS1lbk8e8nMMZmeS2Smh3DakQtkdY4_nonce_': 'iAWZwBXzCs8iVj6l4A6N'}>
```

That way, when the user is redirected back to the `redirect_uri`, `authorize_access_token` would check for both keys, using the value of `state` from `request.args` (i.e., the query string) and only trigger a `mismatching_state` error if either or both are missing?

Many thanks!

**Comment by lepture** [comment]
@dmalan With your fix, there will be a problem for cookie based session. Because the session key is so random, we can't clear the session properly. The cookie string will be larger and larger.

1. End user visit the login url
2. server will set some data into session and then redirect to GitHub, Twitter, Google etc.
3. End user just closes the tab, he doesn't want to authorize
4. those data will be in session without cleanup

**Comment by lepture** [comment]
@dmalan I would suggest you create a blank login page, all `login required` pages will be redirect to this login page.

1. Anonymous user visit `/docs/protected-page`
2. Redirect to `/account/login`
3. On `/account/login` there is a button, click this button to login with third party.

**Comment by dmalan** [comment]
> @dmalan With your fix, there will be a problem for cookie based session. Because the session key is so random, we can't clear the session properly. The cookie string will be larger and larger.
> 
> 1. End user visit the login url
> 2. server will set some data into session and then redirect to GitHub, Twitter, Google etc.
> 3. End user just closes the tab, he doesn't want to authorize
> 4. those data will be in session without cleanup

Definitely possible, though probably not a common case? And sounds like a job for garbage collection, no? Old keys (prefixed with, e.g., `_cs50_authlib_`) in `session` could simply be deleted after N minutes, triggered by subsequent calls into the library?

> @dmalan I would suggest you create a blank login page, all login required pages will be redirect to this login page.
> 
> 1. Anonymous user visit /docs/protected-page
> 1. Redirect to /account/login
> 1. On /account/login there is a button, click this button to login with third party. 

This might indeed decrease the probability of state being clobbered, since a user would now have to click a button in multiple tabs. But the risk would still remain, and that definitely feels like a net negative UX-wise?

Thank you for considering!

**Comment by lepture** [comment]
I think the best way to deal with such problem is adding a cache system. We have a `cache` configure for OAuth 1.0: https://docs.authlib.org/en/latest/client/flask.html#using-cache-for-temporary-credential

I'll refactor the code, if the client has cache system, it will use cache system, if not it will use the cookie session. I think this can solve the problem.

**Comment by TimotheeJeannin** [comment]
Hi,

I also have the error `CSRF Warning! State not equal in request and response.` in my logs multiple times per day. 
It's been a long time I'm searching for the root cause of this issue and I was unable to reproduce until now. Thanks to @dmalan for the investigation! :+1: I'm able to reproduce now with two tabs performing a Google authentication flow in a parallel.

From the logs of my website, (a [screenshot api called ApiFlash](https://apiflash.com/)) I made some statistics and it appears that, during the last 10 days, **6,1% of all Google authentication attempts** failed because of this. I don't really know why users are opening so many tabs and starting multiple authentication flow. Maybe they start the flow and want to check a few more things on the website before proceeding? Maybe they have multiple Google accounts and want to switch? Maybe they are not sure if the `Signup button` and the `Get started` button is the same thing? I don't know. 

If there is anything I can do to help solve this issue, let me know!

**Comment by TimotheeJeannin** [comment]
Also, I think it would be great if a fix for this was based on session. It's not nice to have to setup a cache just to fix a bug.

**Comment by lepture** [comment]
This has been fixed in master branch. It will be released in v1.0.

**Comment by dmalan** [comment]
This is terrific, @lepture. It's working great on our end using `master`. Thanks so much!

**Comment by TimotheeJeannin** [comment]
Thank you very much for the fix @lepture. Do you know when you plan to release this?
